openml/extensions/extension_interface.py

```diff
@@ -0,0 +1,91 @@
+from abc import ABCMeta, abstractmethod
+from typing import Any, List, Optional, Tuple
+
+from openml import OpenMLTask, OpenMLFlow
+from openml.runs.trace import OpenMLTraceIteration, OpenMLRunTrace
+
+
+class Extension:
+    __metaclass__ = ABCMeta
+
+    """Defines the interface to connect machine learning libraries to OpenML-Python.
+    """
+
+    ################################################################################################
+    # Abstract methods for flow serialization and de-serialization
+
+    @abstractmethod
+    def flow_to_model(self, flow: OpenMLFlow) -> Any:
+        pass
+
+    @abstractmethod
+    def model_to_flow(self, model: Any) -> OpenMLFlow:
+        pass
+
+    @abstractmethod
+    def flow_to_parameters(self, model: Any) -> List:
+        pass
+
+    @abstractmethod
+    def is_estimator(self, model: Any) -> bool:
+        pass
+
+    @abstractmethod
+    def seed_model(self, model: Any, seed: Optional[int]) -> Any:
+        pass
+
+    @abstractmethod
+    def run_model_on_fold(
+        self,
+        model: Any,
+        task: OpenMLTask,
+        rep_no: int,
+        fold_no: int,
+        sample_no: int,
+        can_measure_runtime: bool,
+        add_local_measures: bool,
+        extension: 'Extension',
+    ) -> Tuple:
+        pass
+
+    ################################################################################################
+    # Abstract methods for hyperparameter optimization
+
+    @abstractmethod
+    def is_hpo_class(self, model: Any) -> bool:
+        pass
+
+    @abstractmethod
+    def assert_hpo_class(self, model: Any) -> None:
```

Comment:
Should this method be abstract? Leaving it non-abstract makes more sense to me. Perhaps create an else clause which raises NotImplementedError to make sure that any instantiated child has different behavior depending on whether it's a hpo class.

Reply:
What does this method do? (I don't understand the assert part)

Reply:
The assert part reveals the intention that rather than just indicating whether or not it is a hyperparameter optimization class, it should actually raise an error if it isn't. This can be helpful for a piece of code that uses this object *needs/expects* it to be hpo_class, it can first assert that it is and then continue. If it turns out it isn't a hpo_class, the error will be raised which should lead to a quick and clear fail (matches with a fail-early, fail-loudly philosophy which tends to help identify bugs).